kategory-core/src/main/kotlin/kategory/data/Weak.kt

```diff
@@ -0,0 +1,58 @@
+package kategory
+
+import java.lang.ref.WeakReference
+
+/**
+ * Represents an object that can stop existing when no references to it are present. It is backed by
+ * a [WeakReference] instance.
+ */
+@higherkind
+@deriving(
+    Functor::class,
+    Applicative::class,
+    Monad::class,
+    Foldable::class,
+    Traverse::class,
+    TraverseFilter::class,
+    MonadFilter::class)
+data class Weak<out A>(internal val provider: () -> A?) : WeakKind<A> {
+
+    companion object {
+
+        private val EMPTY: Weak<Nothing> = Weak({ null })
+
+        @Suppress("UNCHECKED_CAST")
+        fun <B> emptyWeak(): Weak<B> = EMPTY
+
+        operator fun <A> invoke(a: A): Weak<A> {
+            val reference = WeakReference(a)
+            return Weak { reference.get() }
+        }
+    }
+
+    fun <B> fold(fn: () -> B, f: (A) -> B): B
+        = provider()?.let(f) ?: fn()
```

Comment:
What happens if f returns a null? Does it still call fn()?

Reply:
True, B should extend `Any` Do'h!

Reply:
No, no, it could still be `Any?`. Just use if instead of ternary operator.

Reply:
Ah! `provider()` is private so can't inline (may expose it, not sure). I'll use a `when` similarly to `Option`